enhancements/network/multi-network-vrfs.md

```diff
@@ -0,0 +1,499 @@
+
+
+### Drawbacks
+
+As described in the Design Details section, this proposal will require reserving two IPs per network in the masquerade subnet. This is a private subnet only used internally by OVN-Kubernetes, but it will require increasing the subnet size in order to accommodate multiple networks. Today this subnet by default is configured as a /29 for IPv4, and only 6 IP addresses are used. With this new design, users will need to reconfigure their subnet to be large enough to hold the desired number of networks. Note, it is already supported  to change the masquerade subnet post-installation.
+
+
+## Design Details
+
+OVN offers the ability to create multiple virtual topologies. As with secondary networks in OVN-Kubernetes today, separate topologies are created whenever a new network is needed. The same methodology will be leveraged for this design. Whenever a new network of type layer 3 or layer 2 is requested, a new topology will be created for that network where pods may connect to.
+
+The limitation today with secondary networks is that there is only support for east/west traffic. This RFE will address adding support for unique primary and secondary network north/south support. In order to support north/south traffic, pods on different networks need to be able to egress, typically using the host’s IP. Today in shared gateway mode we use a Gateway Router (GR) in order to provide this external connectivity, while in local gateway mode, the host kernel handles SNAT’ing and routing out egress traffic. Ingress traffic also follows similar, and reverse paths. There are some exceptions to these rules:
+
+
+
+1. MEG traffic always uses the GR to send and receive traffic.
+2. Egress IP on the primary NIC always uses the GR, even in local gateway mode.
+3. Egress Services always use the host kernel for egress routing.
+
+To provide an ingress/egress point for pods on different networks the most simple solution may appear to be to connect them all to a single gateway router. This introduces an issue where now networks are all connected to a single router, and there may be routing happening between networks that were supposed to be isolated from one another. Furthermore in the future, we will want to extend these networks beyond the cluster, and to do that in OVN would require making a single router VRF aware, which adds more complexity into OVN.
+
+The proposal here is to create a GR per network. With this topology, OVN will create a patch port per network to the br-ex bridge. OVN-Kubernetes will be responsible for being VRF/network aware and forwarding packets via flows in br-ex to the right GR. Each per-network GR will only have load balancers configured on it for its network, and only be able to route to pods in its network. The logical topology would look something like this, if we use an example of having a default primary network, a layer 3 primary network, and a layer 2 primary network:
```

Comment:
IIUC all "unique" networks + cluster default network will bind to br-ex. I also assume we won't be using VLANs here. Doesn't this break network isolation in the host ?

Will we have more custom openflow on br-ex to implement traffic isolation between networks ?

Reply:
Right there will be isolation flows in br-ex to ensure that a network cannot talk to another via br-ex switching. Additionally, packets coming out of the network into the mp0 port will isolated into their own VRF.